elliptic-curve/src/arithmetic.rs

```diff
@@ -84,3 +84,27 @@ pub trait PrimeCurveArithmetic:
     /// Prime order elliptic curve group.
     type CurveGroup: group::prime::PrimeCurve<Affine = <Self as CurveArithmetic>::AffinePoint>;
 }
+
+/// Normalize point(s) in projective representation by converting them to their affine ones.
+pub trait Normalize: Sized {
+    /// Elliptic curve point in affine coordinates.
+    type AffinePoint;
+
+    /// Converts the point to its affine representation.
+    fn to_affine(&self) -> Self::AffinePoint;
+
```

Comment:
The associated type for an affine point (`AffineRepr`) and `to_affine` are already provided by [`group::Curve`](https://github.com/RustCrypto/formats/pull/810/files) which is already impl'd for all of the `ProjectivePoint` types.

```suggestion
/// Normalize point(s) in projective representation by converting them to their affine ones.
pub trait Normalize: group::Curve {
```

It also provides a [`batch_normalize`](https://docs.rs/group/latest/group/trait.Curve.html#method.batch_normalize) function, albeit with a slightly different API.

I was suggesting leveraging this trait, rather than duplicating its functionality. It should be possible to conditionally impl it when the `alloc` feature is available, falling back to its default implementation if it isn't available.

Reply:
I know it's provided. But I wanted to have a single trait that holds it all. 

I'm not sure I understand your  requested then. Let's decide on something and merge it, the naunces are becoming a hardel at this point imo. Should we just remove to_affine() and leave everything else as is?

Reply:
I'd like to avoid duplication of functionality, and especially having two methods with the same name that do the same thing defined on two different traits with different type signatures.

If you bound on `group::Curve`, you can use `group::Curve::AffineRepr` instead of having to redefine a separate associated type for that.

Reply:
Ok but still we should expose both the array based and alloc based api ourselves, and also implement the group one?

Reply:
Though I do like your API better than `group::Curve::batch_normalize`, I think the best way to proceed would be to implement that, add a new trait for `batch_normalize_array` (with `group::Curve` as a supertrait), and try to change the upstream API in `group` in the next breaking release.

I also opened an issue for having a similar trait to `batch_normalize_array` upstream in `group` as well: https://github.com/zkcrypto/group/issues/50

If you'd really like to have something like `batch_normalize` but with your original API, I guess that's okay, but I'd suggest giving it a different name to prevent confusion (though I can't think of a good one offhand).

Reply:
But I don't even understand how I can implement `Cruve`'s `batch_normalize`? The batch inversion algorithm requires allocations to be made, and that's the reason I didn't go with this interface in the first place.

I understand that you don't want to have two conflicting APIs, but their API cannot be made to work. Therefore, I suggest to ignore it, and "override" it with ours. 

I now pushed a commit that takes `AffineRepr` from `Curve` though; I think this code should be approved.

Reply:
> But I don't even understand how I can implement Cruve's batch_normalize? The batch inversion algorithm > requires allocations to be made

You can gate your implementation on `#[cfg(feature = "alloc")]`. It's a provided method, so when you don't provide an optimized version it will fall back onto the default implementation, which just loops over the inputs calling `to_affine`.

Reply:
I did as you requested, but this really became a little cumbersome; the conditional implementation doesn't work as swiftly as I'd hope.

> The associated type for an affine point (`AffineRepr`) and `to_affine` are already provided by [`group::Curve`](https://github.com/RustCrypto/formats/pull/810/files) which is [already impl'd for all of the `ProjectivePoint` types](https://docs.rs/k256/latest/k256/struct.ProjectivePoint.html#impl-Curve-for-ProjectivePoint).
> 
> It also provides a [`batch_normalize`](https://docs.rs/group/latest/group/trait.Curve.html#method.batch_normalize) function, albeit with a slightly different API.
> 
> I was suggesting leveraging this trait, rather than duplicating its functionality. It should be possible to conditionally impl `group::Curve::batch_normalize` when the `alloc` feature is available, falling back to its default implementation if it isn't available.

Please see https://github.com/RustCrypto/elliptic-curves/pull/971 and let me know if this is acceptable

Reply:
Fixed